Demo/CqlSdkUseCases/Program.cs

```diff
@@ -0,0 +1,154 @@
+    }
+
+    public LibrarySet LibrarySet { get; }
+
+    public Stream CompileToAssembly(Library library, Stream csharpStream)
+    {
+        var assemblyData = _assemblyCompiler.CompileNode(csharpStream, _results, LibrarySet, library);
+        _results.Add(library.GetVersionedIdentifier()!, assemblyData);
+        return new MemoryStream(assemblyData.Binary);
+    }
+}
+
+
+internal class CSharpToBinaryFactory(
+    AssemblyCompiler assemblyCompiler)
+{
+    public CSharpToBinary ForLibrarySet(LibrarySet librarySet) =>
+        new CSharpToBinary(assemblyCompiler, librarySet);
+}
+
+public class ElmToCSharp
+{
+    private readonly LibrarySetExpressionBuilder _librarySetExpressionBuilder;
+    private readonly CSharpLibrarySetToStreamsWriter _cSharpLibrarySetToStreamsWriter;
+    private readonly LibrarySet _librarySet;
+
+    internal ElmToCSharp(
+        LibrarySetExpressionBuilder librarySetExpressionBuilder,
+        CSharpLibrarySetToStreamsWriter cSharpLibrarySetToStreamsWriter,
+        LibrarySet librarySet)
+    {
+        _librarySetExpressionBuilder = librarySetExpressionBuilder;
+        _cSharpLibrarySetToStreamsWriter = cSharpLibrarySetToStreamsWriter;
+        _librarySet = librarySet;
+    }
+
+    public async IAsyncEnumerable<(Library library, Stream csharpCodeStream)> GenerateCSharp(CancellationToken ct)
+    {
+        ct.ThrowIfCancellationRequested();
+
+        TaskCompletionSource<CSharpSourceCodeStep> tcs = new();
+        Queue<Task<CSharpSourceCodeStep>> steps = new();
+        steps.Enqueue(tcs.Task);
+
+        var definitions = _librarySetExpressionBuilder.ProcessLibrarySet(_librarySet);
+        var callbacks = new CSharpSourceCodeWriterCallbacks(
+            libraryNameToClassName: null,
+            shouldWriteLibrary: null,
+            onAfterStep: OnAfterStep);
+        var processTask = Task.Run(() => _cSharpLibrarySetToStreamsWriter.ProcessDefinitions(definitions, _librarySet, callbacks), ct);
+
+        void OnAfterStep(CSharpSourceCodeStep step)
+        {
+            switch (step)
+            {
+                case CSharpSourceCodeStep.OnStream onStream:
+                    TaskCompletionSource<CSharpSourceCodeStep> tcsNext = new();
+                    steps.Enqueue(tcsNext.Task);
+                    tcs.SetResult(onStream);
+                    tcs = tcsNext;
+                    break;
+                case CSharpSourceCodeStep.OnDone onDone:
+                    tcs.SetResult(onDone);
+                    break;
+            }
+        }
+
+        while (steps.TryDequeue(out var stepTask))
+        {
+            var step = await stepTask.WaitAsync(ct);
+            switch (step)
+            {
+                case OnStream onStream:
+                    yield return (_librarySet.GetLibrary(onStream.Name, true)!, onStream.Stream);
+                    break;
+                case OnDone _:
+                    break;
+            }
+        }
+
+        await processTask;
```

Comment:
I've never seen an async method implemented like this before.  I can pinpoint a couple things.  For example, this function should not be async and you shouldn't `await processTask` on line 143 and just `return processTask` instead.  This lets the caller do the same kind of shenanigans you're doing with tasks here, e.g., they could combine the returned `Task` from this method with other tasks and then use `Task.WaitAll()` on them.  I think this is just _wrong_, but I am having a hard time articulating exactly why, except that I've never seen code that looks anything like this, and that's a signal to me that something is wrong.

I think we need to take this back to the drawing board.  The `CSharpSourceCodeWriterCallbacks` concept is overengineered and I think this method here proves that point.  Look at how complicated this code is.

We should discuss how to rescue the assembly generator.  The code is inherently complicated and requires these steps, but there's a default set of steps that everyone will use 99% of the time.  We should perhaps consider an `event` pattern here.

Also, nobody *ever* uses `CancellationToken`, and certainly not as a mandatory parameter.  Nobody will ever cancel this task.

Reply:
This PR was in draft and not nearly ready for a technical review yet.